planner/cascades/transformation_rules.go

```diff
@@ -710,6 +711,51 @@ func (r *PushLimitDownProjection) OnTransform(old *memo.ExprIter) (newExprs []*m
 	return []*memo.GroupExpr{projExpr}, true, false, nil
 }
 
+// PushLimitDownUnionAll pushes limit to union all.
+type PushLimitDownUnionAll struct {
+	baseRule
+}
+
+// NewRulePushLimitDownUnionAll creates a new Transformation PushLimitDownUnionAll.
+// The pattern of this rule is `Limit->UnionAll->X`.
+func NewRulePushLimitDownUnionAll() Transformation {
+	rule := &PushLimitDownUnionAll{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandLimit,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandUnionAll, memo.EngineTiDBOnly),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+// Use appliedRuleSet in GroupExpr to avoid re-apply rules.
+func (r *PushLimitDownUnionAll) Match(expr *memo.ExprIter) bool {
+	return !expr.GetExpr().HasAppliedRule(r)
+}
+
+// OnTransform implements Transformation interface.
+// It will transform `Limit->UnionAll->X` to `Limit->UnionAll->Limit->X`.
+func (r *PushLimitDownUnionAll) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	limit := old.GetExpr().ExprNode.(*plannercore.LogicalLimit)
+	unionAll := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalUnionAll)
+
+	newUnionAllExpr := memo.NewGroupExpr(unionAll)
+	for _, childGroup := range old.Children[0].GetExpr().Children {
+		newLimitExpr := memo.NewGroupExpr(limit)
```

Comment:
Similar to `PushTopNDownUnionAll`, the new Limit should be: `LogicalLimit{Count: oldLimit.Count + oldLimit.Offset}`. For example:
```
mysql> explain (select a from t) union all (select b from t) limit 3 offset 5;
+------------------------------+-------+-----------+---------------------------------------------------------------------+
| id                           | count | task      | operator info                                                       |
+------------------------------+-------+-----------+---------------------------------------------------------------------+
| Limit_17                     | 3.00  | root      | offset:5, count:3                                                   |
| └─Union_18                   | 8.00  | root      |                                                                     |
|   ├─Limit_20                 | 8.00  | root      | offset:0, count:8                                                   |
|   │ └─TableReader_25         | 8.00  | root      | data:Limit_24                                                       |
|   │   └─Limit_24             | 8.00  | cop[tikv] | offset:0, count:8                                                   |
|   │     └─TableScan_23       | 8.00  | cop[tikv] | table:t, range:[-inf,+inf], keep order:false, stats:pseudo          |
|   └─Limit_29                 | 8.00  | root      | offset:0, count:8                                                   |
|     └─IndexReader_35         | 8.00  | root      | index:Limit_34                                                      |
|       └─Limit_34             | 8.00  | cop[tikv] | offset:0, count:8                                                   |
|         └─IndexScan_33       | 8.00  | cop[tikv] | table:t, index:b, range:[NULL,+inf], keep order:false, stats:pseudo |
+------------------------------+-------+-----------+---------------------------------------------------------------------+
10 rows in set (0.00 sec)
```

Reply:
It makes sense. Bit what I don't understand about your comments is that the logicalLimit we got from the old-expression has consisted of  both Count and offSet, is it necessary to construct a new LogicalLimit from the old one? In my understanding, they're the same.

Reply:
They are not the same, actually. See the example above, the new one should be `offset = 0, count = old.count + old.offset`.

Reply:
got you, will fix.